DR-2436: Snapshot Query

snf2ye opened the pull request:
Changes deployed here:
https://jade-sh.datarepo-dev.broadinstitute.org/

snf2ye committed:
Convert DatasetQueryView to functional component

snf2ye committed:
Move to common folder

snf2ye committed:
Move to common folder

snf2ye committed:
Move query view UI code into common

snf2ye committed:
Point snapshot query view to common ui

snf2ye committed:
query page - query mapped through but getting 404

fix

snf2ye committed:
query page - query mapped through but getting 404

fix

snf2ye committed:
snapshot tables don't need to be prefixed with datarepo_

snf2ye committed:
Convert JadeTable to functional component

snf2ye committed:
Prevent sort for snapshots, convert to functional, work to standardize query objects

snf2ye committed:
Prevent sort for snapshots, convert to functional, work to standardize query objects

snf2ye committed:
Ok, well it loosely "works"

snf2ye committed:
Ok, well it loosely "works"

snf2ye committed:
map in data repo row id

snf2ye committed:
map in data repo row id

snf2ye committed:
sorta get page working

snf2ye committed:
sorta get page working

snf2ye committed:
keep page numbers intact when next page is loading

snf2ye committed:
keep page numbers intact when next page is loading

snf2ye committed:
wip move page and rowsPerPage to state

snf2ye committed:
wip move page and rowsPerPage to state

snf2ye committed:
remove rouge reference to BQ querying

snf2ye committed:
remove rouge reference to BQ querying

snf2ye committed:
cleanup

snf2ye committed:
fix bugs with Michael's help

snf2ye committed:
update data on page change

snf2ye committed:
update data on page change

snf2ye committed:
turn off panels until next PR

snf2ye committed:
Fix dataset view

snf2ye committed:
Fix dataset view

snf2ye committed:
common loading spinner

snf2ye committed:
common loading spinner

snf2ye committed:
add react-hooks rules

snf2ye committed:
Fixed react dependency array issues for datasets

snf2ye committed:
Fixed react dependency array issues for datasets

snf2ye committed:
cleanup unused vars

snf2ye committed:
cleanup unused vars

snf2ye committed:
use paging directly in state rather than passing it around

snf2ye committed:
use paging directly in state rather than passing it around

snf2ye committed:
Fix more lint errors

snf2ye committed:
test if fixes cypress tests

snf2ye committed:
rearrange

snf2ye committed:
Attempt to undo moving files out of /dataset for panel since not yet using

snf2ye committed:
Attempt to undo moving files out of /dataset for panel since not yet using

snf2ye committed:
remove unnecessary change to bigQuery.js since now using preview

fix bug with last change

missing update to reference

snf2ye committed:
remove unnecessary change to bigQuery.js since now using preview

fix bug with last change

missing update to reference

snf2ye committed:
Switch QueryViewDrop to be functional component

snf2ye committed:
Switch QueryViewDrop to be functional component

snf2ye committed:
Attempting to clear up some of the linting errors

snf2ye committed:
moving sorting to state -- Shouldn't update something both in state and store

snf2ye committed:
moving sorting to state -- Shouldn't update something both in state and store

snf2ye committed:
fix lint errors

snf2ye committed:
more linting

nmalfroy commented on src/actions/app.js:
```diff
@@ -168,19 +168,30 @@ export const { hideAlert } = createActions({
+});
+
+export const { previewData } = createActions({
+  [ActionTypes.PREVIEW_DATA_SUCCESS]: (queryResults, columns) => ({
```
minor: could we swap the order?  I like `FOO_SUCCESS` being after `FOO`

nmalfroy replied:
should we also have a FAILURE handler?

snf2ye replied:
Good point - added. I added a "Failure" component as a sorta placeholder -- probably should come up with a design for handling the case when the request fails. Right now it's just an error icon + error message.

nmalfroy commented on src/components/table/JadeTable.jsx:
```diff
@@ -150,82 +106,91 @@ export class JadeTable extends React.PureComponent {
+        {polling && (
+          <LoadingSpinner
+            delay={delay}
+            delayMessage="For large datasets, it can take a few minutes to fetch results from BigQuery. Thank you for your patience."
```
While we're in here, let's maybe take out `from BigQuery` since we're gonna want this for parquet too

nmalfroy commented on src/reducers/query.js:
```diff
@@ -26,13 +32,55 @@ export default {
+          return res;
+        });
+        const queryParams = {
+          // pageToken: queryResults.pageToken,
```
commented code

nmalfroy reviewed (commented):
Did an initial pass. looking good!  I need to click test too

snf2ye committed:
code review changes

snf2ye committed:
bad merge

snf2ye committed:
another fix to bad merge

snf2ye committed:
lint

reorder

making prettier happy

snf2ye committed:
Update error message for test

snf2ye committed:
Step towards a simplier data model for preview endpoints

TODO: figure out where those extra properties are set on the columns and fix homepage

Bug fixes

snf2ye committed:
special case allow sort for main table

snf2ye committed:
move sort to column model

snf2ye committed:
reset query on switching pages

snf2ye committed:
reset query on page change

snf2ye committed:
Fix broken panel

snf2ye committed:
remove unused var

snf2ye committed:
handle merge conflicts

snf2ye committed:
rename extensions

snf2ye committed:
reset query for snapshot too

tlangs commented on src/actions/app.js:
```diff
@@ -168,19 +168,31 @@ export const { hideAlert } = createActions({
 });
 
+export const { previewData } = createActions({
+  [ActionTypes.PREVIEW_DATA]: (snapshotId, table, columns, totalRowCount) => ({
```
Will we be using the preview API for datasets as well as snapshots? If so, we might want to not call the id `snapshotId` here.

tlangs commented on src/actions/app.js:
```diff
@@ -200,6 +212,18 @@ export const { applySort } = createActions({
   }),
 });
 
+export const { resetQuery } = createActions({
```
If these new actions are a part of the new preview API, shouldn't they be declared under the `previewData` constant?

snf2ye replied:
Any of the "query" actions apply to all of the tables, so it wouldn't be appropriate to move this under previewData.

tlangs commented on src/components/common/query/QueryView.jsx:
```diff
@@ -0,0 +1,129 @@
+import { Button, Typography } from '@material-ui/core';
+
+import { Link } from 'react-router-dom';
+import SidebarDrawer from 'components/dataset/data/sidebar/SidebarDrawer';
```
the `SidebarDrawer` should be moved out of the `dataset` folder if its used for both datasets and snapshots.

snf2ye replied:
Agreed - I think this will be covered in https://broadworkbench.atlassian.net/browse/DR-2483 - I'll add this to A.C.

tlangs commented on src/components/common/query/QueryView.jsx:
```diff
@@ -0,0 +1,129 @@
+        <div className={classes.pageRoot}>
+          <AppBreadcrumbs
+            context={{
+              type:
```
I would love to make typescript types for datasets and snapshots. Then, we could just pass the dataset/snapshot to `AppBreadcrumbs` and it could figure out what to do based on the type instead of needing to pass the resource type in as a parameter.

tlangs commented on src/components/common/query/QueryView.jsx:
```diff
@@ -0,0 +1,129 @@
+import React, { Fragment } from 'react';
```
Since this is now referred to as the "Data" page for datasets/snapshots, can we rename files to match?

snf2ye replied:
Hopefully with my last pass, I've changed more of the components to the new names.

tlangs commented on src/components/dataset/data/sidebar/QueryViewSidebarItem.jsx:
```diff
@@ -42,19 +42,19 @@
~    // enable the button when there are unsaved changes
~    if (!_.isEqual(prevState.filterMap, filterMap)) {
~      this.setState({ disableButton: this.invalidChange(filterMap) });
~    }
```
Nice catch!

tlangs commented on src/components/dataset/data/DatasetDataView.jsx:
```diff
@@ -0,0 +1,220 @@
+
+const QUERY_LIMIT = 1000;
+
+function DatasetQueryView({
```
Did we decide to make all new components in Typescript? I can't remember. It would be cool to see Typescript proliferate in our codebase, though!

snf2ye replied:
Definitely - It was just more than I wanted to take on with this PR. I did attempt to rework any models to consistent var names/etc, which should help with that transition.

tlangs commented on src/components/table/DatasetSnapshotsTable.jsx:
```diff
@@ -39,20 +39,23 @@ class DatasetSnapshotsTable extends React.PureComponent {
+        allowSort: true,
         render: (row) => (
-          <Link to={`/snapshots/details/${row.id}`} className={classes.jadeLink}>
+          <Link to={`/snapshots/${row.id}/overview`} className={classes.jadeLink}>
```
There's no more `overview` route. The snapshot overview pages now live at `/snapshot/{ID}`

snf2ye replied:
Oh thanks, I thought I had caught all of the route changes.

tlangs commented on src/components/table/JadeTable.jsx:
```diff
@@ -45,201 +36,172 @@ const styles = (theme) => ({
+                  return (
+                    <TableRow hover tabIndex={-1} key={drId}>
+                      {columns.map((column) => {
+                        const value = handleValues(row[column.name], column);
```
I like this change! Much clearer!

tlangs commented on src/components/DetailViewHeader.jsx:
```diff
@@ -57,6 +58,10 @@ const styles = (theme) => ({
+  },
 });
 
 export class DetailViewHeader extends React.PureComponent {
```
Should this now be `OverviewHeader`?

snf2ye replied:
Yeah - I've done another pass now and renamed a bunch more files to corresponding data/overview names.

tlangs commented on src/reducers/query.js:
```diff
@@ -1,20 +1,28 @@
+import { ActionTypes, TABLE_DEFAULT_ROWS_PER_PAGE } from 'constants/index';
 import BigQuery from 'modules/bigquery';
 
 export const queryState = {
```
I'm wondering if its worth renaming this to `previewState`, since it's using the preview API

snf2ye replied:
There's so much tied into this "query" name - I think we can push off that name migration to when we ditch the BQ query aspect altogether. At the very least, I'd want to punt to another ticket/PR.

tlangs reviewed (changes requested):
This was a hefty effort to take on. Thanks so much for chewing through it!

snf2ye committed:
code review changes - rename components + fix route

snf2ye committed:
code review - make previewData

snf2ye committed:
Fix cypress test

snf2ye committed:
Changes discussed in code review meeting

snf2ye committed:
linter

snf2ye committed:
merge w/ develop

nmalfroy commented on src/components/common/data/DataView.jsx:
```diff
@@ -0,0 +1,129 @@
+  selectedTable: PropTypes.object,
+  sidebarWidth: PropTypes.number,
+  tableNames: PropTypes.array,
+  pageBQQuery: PropTypes.func,
```
Mind just making this alphabetical (or should we get rid of this linting rule?)

nmalfroy commented on src/components/dataset/data/DatasetDataView.jsx:
```diff
@@ -0,0 +1,220 @@
+  const [panels, setPanels] = useState([]);
+
+  useEffect(() => {
+    const datasetId = match.params.uuid;
```
We may want to make the datasetId be a property of the component and do the mapping in a `connect` wrapper.

snf2ye replied:
I'm not sure how to do that. We are already pulling the match value via the connect wrapper.

There is, however,  a hooks implementation we could use instead: https://v5.reactrouter.com/web/api/Hooks/useparams

snf2ye replied:
It also seems like keeping this particular info out of the store makes sense for readability - It's clear that we're triggering updates starting from this component when the url uuid updates.

nmalfroy replied:
I guess I was thinking of doing this mapping explicitly in the mapStateToProps method?  So at the bottom of the file you might have:

```function mapStateToProps(state) {
  return {
    datasets: state.datasets.datasets,
    datasetsCount: state.datasets.datasetsCount,
    filteredDatasetsCount: state.datasets.filteredDatasetsCount,
    features: state.user.features,
    datasetId: state.params.uuid,
  };
}``

And then datasetId is a first class prop of the component.

Basically, consolidate the mapping there...does that make sense?

snf2ye replied:
Yes, that makes sense. 
I'm not sure how to populate the state.params value, though. As far as I can tell, the router state in the store doesn't currently include that information.

nmalfroy replied:
talked offline and it's totally fine to leave

nmalfroy commented on src/components/snapshot/data/SnapshotDataView.jsx:
```diff
@@ -0,0 +1,137 @@
+  const [panels, setPanels] = useState([]);
+
+  useEffect(() => {
+    const snapshotId = match.params.uuid;
```
Same comment about grabbing this from state in a connect method

nmalfroy reviewed (approved):
Man, this PR must have been a slog, thanks for working through it!  Is your deployed instance up to date with this code?  I'll do a list quick click test.  Let's merge this as soon as this week's deploy is done

tlangs reviewed (approved):
Once Muscle's comments are resolved, merge this puppy!

snf2ye commented:
>\s your deployed instance up to date with this code? I'll do a list quick click test. Let's merge this as soon as this week's deploy is done

@nmalfroy  - Yes, latest is deployed

snf2ye committed:
lint

nmalfroy commented:
> > Is your deployed instance up to date with this code? I'll do a list quick click test. Let's merge this as soon as this week's deploy is done
> 
> @nmalfroy - Yes, latest is deployed

Cool, thanks!  I've been click testing and it looks great!
